language/diem-vm/parallel-executor/src/proptest_types/types.rs

```diff
@@ -0,0 +1,328 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    errors::{Error, Result},
+    task::{ExecutionStatus, ExecutorTask, Transaction as TransactionType, TransactionInferencer},
+};
+use anyhow::Result as AResult;
+use mvhashmap::MVHashMapView;
+use proptest::{
+    arbitrary::Arbitrary, collection::vec, prelude::*, proptest, sample::Index, strategy::Strategy,
+};
+use proptest_derive::Arbitrary;
+use std::{
+    collections::{HashMap, HashSet},
+    fmt::Debug,
+    hash::Hash,
+    marker::PhantomData,
+};
+
+///////////////////////////////////////////////////////////////////////////
+// Generation of transactions
+///////////////////////////////////////////////////////////////////////////
+
+#[derive(Arbitrary, Debug, Clone)]
+#[proptest(params = "TransactionGenParams")]
+pub struct TransactionGen<V: Arbitrary + Debug + 'static + Clone> {
+    #[proptest(
+        strategy = "vec((any::<Index>(), value_strategy(params.write_keep_rate)), 1..params.possible_write_size)"
+    )]
+    keys_modified: Vec<(Index, Option<V>)>,
+    #[proptest(strategy = "vec(any::<Index>(), 1..params.read_size)")]
+    keys_read: Vec<Index>,
+}
+
+#[derive(Clone, Copy)]
+pub struct TransactionGenParams {
+    pub possible_write_size: usize,
+    pub read_size: usize,
+    pub write_keep_rate: f64,
+}
+
+/// A naive transaction that could be used to test the correctness and throughput of the system.
+#[derive(Debug, Clone)]
+pub enum Transaction<K, V> {
+    Write {
+        /// Write to some keys with value provided.
+        actual_writes: Vec<(K, V)>,
+        /// Skipp writing to some keys. This is used to simulate over approximation of the inferencer.
+        skipped_writes: Vec<K>,
+        /// Read from some keys.
+        reads: Vec<K>,
+    },
+    /// Skip the execution of trailing transactions.
+    SkipRest,
+    /// Abort the execution.
+    Abort,
+}
+
+impl Default for TransactionGenParams {
+    fn default() -> Self {
+        TransactionGenParams {
+            possible_write_size: 10,
+            write_keep_rate: 0.5,
+            read_size: 10,
+        }
+    }
+}
+
+fn value_strategy<V: Arbitrary + Clone + 'static>(
+    keep_rate: f64,
+) -> impl Strategy<Value = Option<V>> {
+    let value_strategy = any::<V>();
+    proptest::option::weighted(keep_rate, value_strategy)
+}
+
+impl<V: Arbitrary + Debug + Clone> TransactionGen<V> {
+    pub fn materialize<K: Clone + Hash + Eq>(self, universe: &[K]) -> Transaction<K, V> {
+        let mut keys_modified = HashSet::new();
```

Comment:
Nit: maybe btreemap in case there could be some level of nondeterminism introduces in the `actual_writes` vec from this